File: horizontal-picker/src/main/java/com/github/jhonnyx2012/horizontalpicker/HorizontalPicker.java
package com.github.jhonnyx2012.horizontalpicker;

import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.support.annotation.ColorInt;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.joda.time.DateTime;

/**
 * Created by Jhonny Barrios on 22/02/2017.
 */

public class HorizontalPicker extends LinearLayout {

    private static final int
            DEFAULT_DAYS = 90,
            DEFAULT_OFFSET = 7;

    private DatePickerListener listener;

    private HorizontalPickerRecyclerView rvDays;

    private View vHover;
    private TextView tvMonth;
    private TextView tvToday;

    // number of generated days
    private int days = -1;

    // number of days before today's date
    private int offset = -1;

    private boolean showTodayButton = true;
    private boolean selectOnScroll = true;

    public HorizontalPicker(Context context) {
        super(context);
    }

    public HorizontalPicker(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

    }

    public HorizontalPicker(Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    public HorizontalPicker setListener(DatePickerListener listener) {
        this.listener = listener;
        return this;
    }

    public void setDate(final DateTime date) {
        rvDays.post(new Runnable() {
            @Override
            public void run() {
                rvDays.setDate(date);
            }
        });
    }

    public void init() {
        inflate(getContext(), R.layout.horizontal_picker, this);
        rvDays = findViewById(R.id.rvDays);

        int finalDays = days == -1 ? DEFAULT_DAYS : days;
        int finalOffset = offset == -1 ? DEFAULT_OFFSET : offset;

        vHover = findViewById(R.id.vHover);
        tvMonth = findViewById(R.id.tvMonth);

        rvDays.setListener(new HorizontalPickerListener() {
            @Override
            public void onStopDraggingPicker() {
                if (vHover.getVisibility() == VISIBLE) vHover.setVisibility(INVISIBLE);
            }

            @Override
            public void onDraggingPicker() {
                if (vHover.getVisibility() == INVISIBLE) vHover.setVisibility(VISIBLE);
            }

            @Override
            public void onDateSelected(Day item) {
                tvMonth.setText(item.getMonth());
                if (showTodayButton)
                    tvToday.setVisibility(item.isToday() ? INVISIBLE : VISIBLE);
                if (listener != null) {
                    listener.onDateSelected(item.getDate());
                }
            }
        });

        tvToday = findViewById(R.id.tvToday);
        tvToday.setOnClickListener(rvDays);
        tvToday.setVisibility(showTodayButton ? VISIBLE : INVISIBLE);
        tvToday.setTextColor(getColor(R.color.colorPrimary));

        int mBackgroundColor = getBackgroundColor();
        setBackgroundColor(mBackgroundColor != Color.TRANSPARENT ? mBackgroundColor : Color.WHITE);

        rvDays.init(
                finalDays,
                finalOffset,
                mBackgroundColor,
                selectOnScroll);
    }

    private int getColor(int colorId) {
        return getResources().getColor(colorId);
    }

    public int getBackgroundColor() {
        int color = Color.TRANSPARENT;
        Drawable background = getBackground();
        if (background instanceof ColorDrawable)
            color = ((ColorDrawable) background).getColor();
        return color;
    }

    @Override
    public boolean post(Runnable action) {
        return rvDays.post(action);
    }

    /**
     * Number of days to be generated in the view. Default value is 90 days.
     * @param days number of days to generate
     * @throws IllegalArgumentException if passed number of days is negative
     * @return this HorizontalPicker instance for chaining method calls
     */
    public HorizontalPicker setDays(int days) {
        if (days < 0) throw new IllegalArgumentException("number of days must be positive");
        this.days = days;
        return this;
    }

    public int getDays() {
        return days;
    }

    /**
     * Number of days to offset from today. Setting offset to 0 will make all but today's day in the
     * future. Default value is 7 days.
     * @param offset number of days before today's date, must be positive
     * @throws IllegalArgumentException if passed offset is negative
     * @return this HorizontalPicker instance for chaining method calls
     */
    public HorizontalPicker setOffset(int offset) {
        if (offset < 0) throw new IllegalArgumentException("offset must be positive");
        this.offset = offset;
        return this;
    }

    public int getOffset() {
        return offset;
    }

    public HorizontalPicker showTodayButton(boolean show) {
        showTodayButton = show;
        return this;
    }

    public HorizontalPicker setSelectOnScroll(boolean selectOnScroll) {
        this.selectOnScroll = selectOnScroll;
        return this;
    }
}

File: horizontal-picker/src/main/kotlin/com/github/mcebular/horizontalpicker/HorizontalPicker.kt
package com.github.mcebular.horizontalpicker

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.LinearLayout
import androidx.core.content.ContextCompat
import com.github.mcebular.horizontalpicker.databinding.HorizontalPickerBinding
import java.time.LocalDate

class HorizontalPicker : LinearLayout {

    companion object {
        private val DEFAULT_DAYS = 90
        private val DEFAULT_OFFSET = 7
    }

    var showTodayButton: Boolean = true
    var selectOnScroll: Boolean = true

    var days: Int = -1
    var offset: Int = -1

    private lateinit var binding: HorizontalPickerBinding
    private var listener: DatePickerListener? = null

    constructor(context: Context?) : super(context)

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    )

    constructor(
        context: Context?,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int
    ) : super(context, attrs, defStyleAttr, defStyleRes)

    override fun post(action: Runnable?): Boolean {
        return binding.recyclerView.post(action)
    }

    fun init() {
        val view = inflate(context, R.layout.horizontal_picker, this)
        binding = HorizontalPickerBinding.bind(view)

        binding.recyclerView.listener = object : HorizontalPickerListener {
            override fun onStopDraggingPicker() {
                if (binding.layoutHover.visibility == VISIBLE) binding.layoutHover.visibility = INVISIBLE
            }

            override fun onDraggingPicker() {
                if (binding.layoutHover.visibility == INVISIBLE) binding.layoutHover.visibility = VISIBLE
            }

            override fun onDateSelected(item: Day) {
                binding.textViewMonth.text = item.month
                if (showTodayButton) binding.textViewToday.visibility = if (item.isToday) INVISIBLE else VISIBLE
                listener?.onDateSelected(item.date)
            }
        }

        binding.textViewToday.let {
            it.setOnClickListener(binding.recyclerView)
            it.visibility = if (showTodayButton) VISIBLE else INVISIBLE
            it.setTextColor(ContextCompat.getColor(context, R.color.colorPrimary))
        }

        val bgColor = getBackgroundColor()
        setBackgroundColor(if (bgColor != Color.TRANSPARENT) bgColor else Color.WHITE)

        binding.recyclerView.init(
            if (days < 0) DEFAULT_DAYS else days,
            if (offset < 0) DEFAULT_OFFSET else offset,
            selectOnScroll
        )
    }

    //

    fun setListener(listener: DatePickerListener): HorizontalPicker {
        this.listener = listener
        return this
    }

    /**
     * Number of days to be generated in the view. Default value is 90 days.
     * @param days number of days to generate
     * @throws IllegalArgumentException if [days] parameter is negative
     * @return this [HorizontalPicker] instance for chaining method calls
     */
    fun setDays(days: Int): HorizontalPicker {
        if (days < 0) throw IllegalArgumentException("Number of days must be positive")
        this.days = days
        return this
    }

    /**
     * Number of days to offset from today. Setting offset to 0 will make all but today's day in the
     * future. Default value is 7 days.
     * @param offset number of days before today's date
     * @throws IllegalArgumentException if [offset] parameter is negative
     * @return this [HorizontalPicker] instance for chaining method calls
     */
    fun setOffset(offset: Int): HorizontalPicker {
        if (offset < 0) throw IllegalArgumentException("Offset must be positive")
        this.offset = offset
        return this
    }

    fun setShowTodayButton(showTodayButton: Boolean): HorizontalPicker {
        this.showTodayButton = showTodayButton
        return this
    }

    fun setSelectOnScroll(selectOnScroll: Boolean): HorizontalPicker {
        this.selectOnScroll = selectOnScroll
        return this
    }

    fun setDate(date: LocalDate) {
        binding.recyclerView.let {
            it.post { it.setDate(date) }
        }
    }

    //

    private fun getBackgroundColor(): Int {
        var color = Color.TRANSPARENT
        val background = background
        if (background is ColorDrawable) color = background.color
        return color
    }

}